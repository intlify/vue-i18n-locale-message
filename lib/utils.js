"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const component_compiler_utils_1 = require("@vue/component-compiler-utils");
const compiler = __importStar(require("vue-template-compiler"));
const fs_1 = __importDefault(require("fs"));
const glob_1 = __importDefault(require("glob"));
const path_1 = __importDefault(require("path"));
const json5_1 = __importDefault(require("json5"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const debug_1 = require("debug");
const debug = debug_1.debug('vue-i18n-locale-message:utils');
const ESC = {
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    '&': '&amp;'
};
function escape(s) {
    return s.replace(/[<>"&]/g, escapeChar);
}
exports.escape = escape;
function escapeChar(a) {
    return ESC[a] || a;
}
function resolve(...paths) {
    return path_1.default.resolve(...paths);
}
exports.resolve = resolve;
function reflectSFCDescriptor(basePath, components) {
    return components.map(target => {
        const { template, script, styles, customBlocks } = component_compiler_utils_1.parse({
            source: target.content,
            filename: target.path,
            compiler: compiler
        });
        return {
            ...parsePath(basePath, target.path),
            raw: target.content,
            customBlocks,
            template,
            script,
            styles
        };
    });
}
exports.reflectSFCDescriptor = reflectSFCDescriptor;
function parsePath(basePath, targetPath) {
    const { dir, name } = path_1.default.parse(targetPath);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [_, target] = dir.split(basePath);
    const parsedTargetPath = target.split(path_1.default.sep);
    parsedTargetPath.shift();
    debug(`parsePath: contentPath = ${targetPath}, component = ${name}, messageHierarchy = ${parsedTargetPath}`);
    return {
        contentPath: targetPath,
        component: name,
        hierarchy: [...parsedTargetPath, name]
    };
}
exports.parsePath = parsePath;
function parseContent(content, lang) {
    switch (lang) {
        case 'yaml':
        case 'yml':
            return js_yaml_1.default.safeLoad(content);
        case 'json5':
            return json5_1.default.parse(content);
        case 'json':
        default:
            return JSON.parse(content);
    }
}
exports.parseContent = parseContent;
function stringifyContent(content, lang, options) {
    var _a, _b;
    const indent = ((_a = options) === null || _a === void 0 ? void 0 : _a.intend) || 2;
    const eof = ((_b = options) === null || _b === void 0 ? void 0 : _b.eof) || '\n';
    let result = '';
    switch (lang) {
        case 'yaml':
        case 'yml':
            result = js_yaml_1.default.safeDump(content, { indent });
            break;
        case 'json5':
            result = json5_1.default.stringify(content, null, indent);
            break;
        case 'json':
        default:
            result = JSON.stringify(content, null, indent);
            break;
    }
    if (!result.endsWith(eof)) {
        result += eof;
    }
    return result;
}
exports.stringifyContent = stringifyContent;
function readSFC(target) {
    const targets = resolveGlob(target);
    debug('readSFC: targets = ', targets);
    // TODO: async implementation
    return targets.map(target => {
        const data = fs_1.default.readFileSync(target);
        return {
            path: target,
            content: data.toString()
        };
    });
}
exports.readSFC = readSFC;
function resolveGlob(target) {
    // TODO: async implementation
    return glob_1.default.sync(`${target}/**/*.vue`);
}
