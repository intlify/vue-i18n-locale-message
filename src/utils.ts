// import types
import { Arguments } from 'yargs'
import { SFCDescriptor } from 'vue-template-compiler'
import { VueTemplateCompiler } from '@vue/component-compiler-utils/dist/types'
import {
  SFCFileInfo,
  Locale,
  LocaleMessageDictionary,
  LocaleMessage,
  LocaleMessages,
  MetaExternalLocaleMessages,
  FormatOptions,
  ProviderFactory,
  ProviderConfiguration,
  TranslationStatusOptions,
  TranslationStatus,
  RawLocaleMessage,
  NamespaceDictionary
} from '../types'

// import modules
import { parse } from '@vue/component-compiler-utils'
import * as compiler from 'vue-template-compiler'
import fs from 'fs'
import glob from 'glob'
import path from 'path'
import JSON5 from 'json5'
import yaml from 'js-yaml'
import deepmerge from 'deepmerge'
import { promisify } from 'util'

import { debug as Debug } from 'debug'
const debug = Debug('vue-i18n-locale-message:utils')

const readFile = promisify(fs.readFile)

// define types
export type PushableOptions = {
  target?: string
  locale?: string
  targetPaths?: string
  filenameMatch?: string
  format?: string
}

const ESC: { [key in string]: string } = {
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  '&': '&amp;'
}

function escapeChar (a: string): string {
  return ESC[a] || a
}

export function escape (s: string): string {
  return s.replace(/[<>"&]/g, escapeChar)
}

export function resolve (...paths: string[]): string {
  return path.resolve(...paths)
}

export function isLocaleMessageDictionary (message: LocaleMessage): message is LocaleMessageDictionary {
  return typeof message !== 'string' && !Array.isArray(message)
}

export function reflectSFCDescriptor (basePath: string, components: SFCFileInfo[]): SFCDescriptor[] {
  return components.map(target => {
    const { template, script, styles, customBlocks } = parse({
      source: target.content,
      filename: target.path,
      compiler: compiler as VueTemplateCompiler
    }) as SFCDescriptor
    return {
      ...parsePath(basePath, target.path),
      raw: target.content,
      customBlocks,
      template,
      script,
      styles
    }
  })
}

export function parsePath (basePath: string, targetPath: string) {
  const { dir, name } = path.parse(targetPath)
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [_, target] = dir.split(basePath)
  const parsedTargetPath = target.split(path.sep)
  parsedTargetPath.shift()
  debug(`parsePath: contentPath = ${targetPath}, component = ${name}, messageHierarchy = ${parsedTargetPath}`)
  return {
    contentPath: targetPath,
    component: name,
    hierarchy: [...parsedTargetPath, name]
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function parseContent (content: string, lang: string): any {
  switch (lang) {
    case 'yaml':
    case 'yml':
      return yaml.safeLoad(content)
    case 'json5':
      return JSON5.parse(content)
    case 'json':
    default:
      return JSON.parse(content)
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function stringifyContent (content: any, lang: string, options?: FormatOptions): string {
  const indent = options?.intend || 2
  const eof = options?.eof || '\n'

  let result = ''
  switch (lang) {
    case 'yaml':
    case 'yml':
      result = yaml.safeDump(content, { indent })
      break
    case 'json5':
      result = JSON5.stringify(content, null, indent)
      break
    case 'json':
    default:
      result = JSON.stringify(content, null, indent)
      break
  }

  if (!result.endsWith(eof)) {
    result += eof
  }

  return result
}

export function readSFC (target: string): SFCFileInfo[] {
  const targets = resolveGlob(target)
  debug('readSFC: targets = ', targets)

  // TODO: async implementation
  return targets.map(target => {
    const data = fs.readFileSync(target)
    return {
      path: target,
      content: data.toString()
    }
  })
}

function resolveGlob (target: string) {
  // TODO: async implementation
  return glob.sync(`${target}/**/*.vue`)
}

export const DEFUALT_CONF = { provider: {}} as ProviderConfiguration

export function resolveProviderConf (provider: string, conf?: string) {
  if (conf) {
    return resolve(conf)
  } else {
    const parsed = path.parse(provider)
    return resolve(process.cwd(), `${parsed.base}-conf.json`)
  }
}

export function loadProvider (provider: string): ProviderFactory | null {
  let mod: ProviderFactory | null = null
  try {
    // TODO: should validate I/F checking & dynamic importing
    const m = require(require.resolve(provider))
    debug('loaderProvider', m)
    if ('__esModule' in m) {
      mod = m.default as ProviderFactory
    } else {
      mod = m as ProviderFactory
    }
  } catch (e) { } // eslint-disable-line
  return mod
}

export function loadProviderConf (confPath: string): ProviderConfiguration {
  let conf = DEFUALT_CONF
  try {
    // TODO: should validate I/F checking & dynamic importing
    conf = require(confPath) as ProviderConfiguration
  } catch (e) { } // eslint-disable-line
  return conf
}

export function getLocaleMessages (args: Arguments<PushableOptions>): LocaleMessages {
  let messages = {} as LocaleMessages

  if (args.target) {
    const targetPath = resolve(args.target)
    const parsed = path.parse(targetPath)
    const locale = args.locale ? args.locale : parsed.name
    messages = Object.assign(messages, { [locale]: require(targetPath) })
  } else if (args.targetPaths) {
    const filenameMatch = args.filenameMatch
    if (!filenameMatch) {
      // TODO: should refactor console message
      throw new Error('You need to specify together --filename-match')
    }
    const targetPaths = args.targetPaths.split(',').filter(p => p)
    targetPaths.forEach(targetPath => {
      const globedPaths = glob.sync(targetPath).map(p => resolve(p))
      globedPaths.forEach(fullPath => {
        const parsed = path.parse(fullPath)
        const re = new RegExp(filenameMatch, 'ig')
        const match = re.exec(parsed.base)
        debug('regex match', match, fullPath)
        if (match && match[1]) {
          const locale = match[1]
          messages = Object.assign(messages, { [locale]: require(fullPath) })
        } else {
          // TODO: should refactor console message
          console.log(`${fullPath} is not matched with ${filenameMatch}`)
        }
      })
    })
  }

  return messages
}

export function getRawLocaleMessages (args: Arguments<PushableOptions>): RawLocaleMessage[] {
  const messages = [] as RawLocaleMessage[]

  if (args.target) {
    const targetPath = resolve(args.target)
    const parsed = path.parse(targetPath)
    const targetFormat = parsed.ext.split('.').pop()
    const format = targetFormat || args.format
    if (!format) {
      // TODO: should refactor console message
      console.log(`ignore ${targetPath}, due to be not specified with --format`)
    } else {
      messages.push({
        locale: args.locale ? args.locale : parsed.name,
        format,
        data: fs.readFileSync(targetPath)
      })
    }
  } else if (args.targetPaths) {
    const filenameMatch = args.filenameMatch
    if (!filenameMatch) {
      // TODO: should refactor console message
      throw new Error('You need to specify together --filename-match')
    }
    const targetPaths = args.targetPaths.split(',').filter(p => p)
    targetPaths.forEach(targetPath => {
      const globedPaths = glob.sync(targetPath).map(p => resolve(p))
      globedPaths.forEach(fullPath => {
        const parsed = path.parse(fullPath)
        const re = new RegExp(filenameMatch, 'ig')
        const match = re.exec(parsed.base)
        debug('regex match', match, fullPath)
        if (match && match[1]) {
          const targetFormat = parsed.ext.split('.').pop()
          const format = targetFormat || args.format
          if (!format) {
            // TODO: should refactor console message
            console.log(`ignore ${fullPath}, due to be not specified with --format`)
          } else {
            messages.push({
              locale: match[1],
              format,
              data: fs.readFileSync(fullPath)
            })
          }
        } else {
          // TODO: should refactor console message
          console.log(`${fullPath} is not matched with ${filenameMatch}`)
        }
      })
    })
  }

  return messages
}

export async function getTranslationStatus (options: TranslationStatusOptions): Promise<TranslationStatus[]> {
  const ProviderFactory = loadProvider(options.provider)
  if (ProviderFactory === null) {
    return Promise.reject(new Error(`Not found ${options.provider} provider`))
  }

  const confPath = resolveProviderConf(options.provider, options.conf)
  const conf = loadProviderConf(confPath) || DEFUALT_CONF

  const locales = options.locales?.split(',').filter(p => p) as Locale[] || []
  const provider = ProviderFactory(conf)
  const status = await provider.status({ locales })
  return Promise.resolve(status)
}

export async function loadNamespaceDictionary (path: string) {
  const raw = await readFile(resolve(path))
  return new Promise<NamespaceDictionary>((resolv, reject) => {
    try {
      // TODO: should be checked more strongly
      resolv(JSON.parse(raw.toString()) as NamespaceDictionary)
    } catch (e) {
      reject(e)
    }
  })
}

type ParsedLocaleMessagePathInfo = {
  locale: Locale
  filename?: string
}

function getLocaleMessagePathInfo (fullPath: string, bundleMatch?: string): ParsedLocaleMessagePathInfo {
  const parsed = path.parse(fullPath)
  debug('getLocaleMessagePathInfo: parsed', parsed)
  if (bundleMatch) {
    const re = new RegExp(bundleMatch, 'ig')
    const match = re.exec(fullPath)
    debug('getLocaleMessagePathInfo: regex match', match)
    return {
      locale: (match && match[1]) ? match[1] : '',
      filename: (match && match[2]) ? match[2] : ''
    }
  } else {
    return {
      locale: parsed.ext.split('.').pop() || ''
    }
  }
}

export function getExternalLocaleMessages (
  dictionary: NamespaceDictionary, bundleWith?: string, bundleMatch?: string
) {
  if (!bundleWith) { return {} }

  const bundleTargetPaths = bundleWith.split(',').filter(p => p)
  return bundleTargetPaths.reduce((messages, targetPath) => {
    const namespace = dictionary[targetPath] || ''
    const globedPaths = glob.sync(targetPath).map(p => resolve(p))
    return globedPaths.reduce((messages, fullPath) => {
      const { locale, filename } = getLocaleMessagePathInfo(fullPath, bundleMatch)
      if (!locale) { return messages }
      const externalMessages = JSON.parse(fs.readFileSync(fullPath).toString())
      let workMessages = externalMessages
      if (filename) {
        workMessages = Object.assign({}, { [filename]: workMessages })
      }
      if (namespace) {
        workMessages = Object.assign({}, { [namespace]: workMessages })
      }
      debug('getExternalLocaleMessages: workMessages', workMessages)
      if (messages[locale]) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        messages[locale] = deepmerge(messages[locale] as any, workMessages)
      } else {
        messages = Object.assign(messages, { [locale]: workMessages })
      }
      debug('getExternalLocaleMessages: messages (processing)', messages)
      return messages
    }, messages)
  }, {} as LocaleMessages)
}

type ExternalLocaleMessagesParseInfo = {
  path: string
  namespace: string
  locale: Locale
  filename?: string
}

// TODO: should be selected more other library ...
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function deepCopy (obj: any): any {
  return JSON.parse(JSON.stringify(obj))
}

export function splitLocaleMessages (
  messages: LocaleMessages,
  dictionary: NamespaceDictionary,
  bundle?: string,
  bundleMatch?: string
) {
  if (!bundle) { return { sfc: messages } }

  const bundleTargetPaths = bundle.split(',').filter(p => p)
  const externalLocaleMessagesParseInfo = bundleTargetPaths.reduce((info, targetPath) => {
    const namespace = dictionary[targetPath] || ''
    const globedPaths = glob.sync(targetPath).map(p => resolve(p))
    debug('splitLocaleMessages globedPaths', globedPaths)
    return globedPaths.reduce((info, fullPath) => {
      const { locale, filename } = getLocaleMessagePathInfo(fullPath, bundleMatch)
      if (!locale) { return info }
      info.push({ path: fullPath, locale, namespace, filename })
      return info
    }, info)
  }, [] as ExternalLocaleMessagesParseInfo[])
  debug('splitLocaleMessages: externalLocaleMessagesParseInfo:', externalLocaleMessagesParseInfo)

  debug('splitLocaleMessages: messages (before):', messages)
  const metaExternalLocaleMessages = externalLocaleMessagesParseInfo.reduce((meta, { path, locale, namespace, filename }) => {
    let targetLocaleMessage = messages[locale]
    if (namespace && isLocaleMessageDictionary(targetLocaleMessage)) {
      targetLocaleMessage = targetLocaleMessage[namespace]
    }
    if (filename && isLocaleMessageDictionary(targetLocaleMessage)) {
      targetLocaleMessage = targetLocaleMessage[filename]
    }
    meta.push({ path, messages: deepCopy(targetLocaleMessage) })
    return meta
  }, [] as MetaExternalLocaleMessages[])
  debug('splitLocaleMessages: messages (after):', messages)
  debug('splitLocaleMessages: metaExternalLocaleMessages:', metaExternalLocaleMessages)

  return { sfc: messages, external: metaExternalLocaleMessages }
}
