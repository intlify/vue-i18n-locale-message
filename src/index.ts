import squeeze from './squeezer'
import infuse from './infuser'
import { getTranslationStatus as status } from './utils'

export {
  squeeze,
  infuse,
  status
}
