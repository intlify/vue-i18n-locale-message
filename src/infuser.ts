/* eslint-disable-next-line */
/// <reference path="../types/shims-vue-template-compiler.d.ts"/>
import { SFCBlock, SFCDescriptor } from 'vue-template-compiler'
import { Locale, MetaLocaleMessage, SFCI18nBlock, SFCFileInfo, FormatOptions } from '../types'

import { escape, reflectSFCDescriptor, parseContent, stringifyContent } from './utils'

import { debug as Debug } from 'debug'
const debug = Debug('vue-i18n-locale-message:infuser')

export default function infuse (basePath: string, sources: SFCFileInfo[], meta: MetaLocaleMessage, options?: FormatOptions): SFCFileInfo[] {
  const descriptors = reflectSFCDescriptor(basePath, sources)

  return descriptors.map(descriptor => {
    return {
      content: generate(meta, descriptor, options),
      path: descriptor.contentPath
    } as SFCFileInfo
  })
}

function generate (meta: MetaLocaleMessage, descriptor: SFCDescriptor, options?: FormatOptions): string {
  const i18nBlocks = meta.components[descriptor.contentPath]
  debug('target i18n blocks\n', i18nBlocks)

  const blocks: SFCBlock[] = getBlocks(descriptor)
  blocks.forEach(b => debug(`block: type=${b.type}, start=${b.start}, end=${b.end}`))

  const { raw } = descriptor
  const content = buildContent(i18nBlocks, raw, blocks, options)
  debug(`build content:\n${content}`)
  debug(`content size: raw=${raw.length}, content=${content.length}`)

  return content
}

function getBlocks (descriptor: SFCDescriptor): SFCBlock[] {
  const { template, script, styles, customBlocks } = descriptor
  const blocks: SFCBlock[] = [...styles, ...customBlocks]
  template && blocks.push(template as SFCBlock)
  script && blocks.push(script as SFCBlock)
  blocks.sort((a, b) => { return (a.start as number) - (b.start as number) })
  return blocks
}

function buildContent (i18nBlocks: SFCI18nBlock[], raw: string, blocks: SFCBlock[], options?: FormatOptions): string {
  let offset = 0
  let i18nBlockCounter = 0
  let contents: string[] = []

  contents = blocks.reduce((contents, block) => {
    if (block.type === 'i18n') {
      let lang = block.attrs.lang
      lang = (!lang || typeof lang !== 'string') ? 'json' : lang
      const locale: Locale | undefined = block.attrs.locale
      const i18nBlock = i18nBlocks[i18nBlockCounter]
      debug(`meta.lang = ${i18nBlock.lang}, block.lang = ${lang}, meta.locale = ${i18nBlock.locale}, block.locale = ${locale}`)

      let messages: any = null
      if (lang === i18nBlock.lang && locale === i18nBlock.locale) {
        if (locale) {
          messages = i18nBlock.messages[locale]
        } else {
          messages = i18nBlock.messages
        }
      } else {
        debug(`unmatch meta block and sfc block`)
        messages = parseContent(block.content, lang)
      }

      contents = contents.concat(raw.slice(offset, block.start))
      const serialized = `\n${stringifyContent(messages, lang, options)}`
      contents = contents.concat(serialized)
      offset = block.end as number
      i18nBlockCounter++
    } else {
      contents = contents.concat(raw.slice(offset, block.end))
      offset = block.end as number
    }
    return contents
  }, contents)
  contents = contents.concat(raw.slice(offset, raw.length))

  if (i18nBlocks.length > i18nBlockCounter) {
    i18nBlocks.slice(i18nBlockCounter).reduce((contents, i18nBlock) => {
      contents.push(buildI18nTag(i18nBlock, options))
      return contents
    }, contents)
  }

  return contents.join('')
}

function buildI18nTag (i18nBlock: SFCI18nBlock, options?: FormatOptions): string {
  const { locale, lang, messages } = i18nBlock
  let tag = '<i18n'
  if (locale) {
    tag += ` locale="${escape(locale)}"`
  }
  if (lang !== 'json') {
    tag += ` lang="${escape(lang)}"`
  }
  tag += '>'

  return `\n
${tag}
${stringifyContent(locale ? messages[locale] : messages, lang, options)}</i18n>`
}
