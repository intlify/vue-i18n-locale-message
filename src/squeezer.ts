import { SFCBlock } from 'vue-template-compiler'
import { MetaLocaleMessage, I18nLang, SFCFileInfo, SFCI18nBlock } from '../types'

import { reflectSFCDescriptor, parseContent } from './utils'

import { debug as Debug } from 'debug'
const debug = Debug('vue-i18n-locale-message:squeezer')

export default function squeeze (basePath: string, files: SFCFileInfo[]): MetaLocaleMessage {
  const descriptors = reflectSFCDescriptor(basePath, files)
  return descriptors.reduce((meta, descriptor) => {
    descriptor.customBlocks.sort((a, b) => { return (a.start as number) - (b.start as number) })
    const i18nBlocks = squeezeFromCustomBlocks(descriptor.customBlocks)
    debug('squeezeFromCustomBlocks: i18nBlocks', JSON.stringify(i18nBlocks, null, 2))
    meta.components[descriptor.contentPath] = i18nBlocks
    return meta
  }, { target: basePath, components: {}} as MetaLocaleMessage)
}

function squeezeFromCustomBlocks (blocks: SFCBlock[]): SFCI18nBlock[] {
  return blocks.map(block => {
    if (block.type === 'i18n') {
      debug('i18n block attrs', block.attrs)

      let lang = block.attrs.lang
      lang = (!lang || typeof lang !== 'string') ? 'json' : lang

      const i18nBlock: SFCI18nBlock = {
        lang: lang as I18nLang,
        messages: {}
      }
      const obj = parseContent(block.content, lang)

      const locale = block.attrs.locale
      if (!locale || typeof locale !== 'string') {
        Object.assign(i18nBlock.messages, obj)
      } else {
        i18nBlock.locale = locale
        Object.assign(i18nBlock.messages, { [locale]: obj })
      }

      return i18nBlock
    }
  }).filter(Boolean) as SFCI18nBlock[]
}
