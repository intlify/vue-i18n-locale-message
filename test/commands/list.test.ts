import * as yargs from 'yargs'
import * as path from 'path'

// -------
// mocking

// mock: ../../src/utils
jest.mock('../../src/utils', () => ({
  __esModule: true,
  ...jest.requireActual('../../src/utils'),
  getLocaleMessages: jest.fn()
}))
import * as utils from '../../src/utils'

jest.mock('fs', () => ({
  ...jest.requireActual('fs'),
  writeFile: jest.fn()
}))
import fs from 'fs'

// -------------------
// setup/teadown hooks

let orgExit // for process.exit mock
let spyLog
let spyError
beforeEach(() => {
  spyLog = jest.spyOn(global.console, 'log')
  spyError = jest.spyOn(global.console, 'error')
  process.exit = jest.fn((code => { return 'exit!' as never })) // mock: process.exit
})

afterEach(() => {
  spyError.mockRestore()
  spyLog.mockRestore()
  jest.clearAllMocks()
  process.exit = orgExit
})

// ----------
// test cases

test('require option', async () => {
  const list = await import('../../src/commands/list')
  const cmd = yargs.command(list)
  try {
    await new Promise((resolve, reject) => {
      cmd.parse(`list`, (err, argv, output) => {
        err ? reject(err) : resolve(output)
      })
    })
  } catch (e) {
    expect(e).toMatchObject({ name: 'YError' })
  }
})

test('not specified --target and --targetPaths', async () => {
  // run
  const list = await import('../../src/commands/list')
  const cmd = yargs.command(list)
  await new Promise((resolve, reject) => {
    cmd.parse(`list --locale=en`, (err, argv, output) => {
      err ? reject(err) : resolve(output)
    })
  })

  // verify
  expect(spyError).toHaveBeenCalledWith('You need to specify either --target or --target-paths')
})

test('--target option', async () => {
  // setup mocks
  const mockUtils = utils as jest.Mocked<typeof utils>
  mockUtils.getLocaleMessages.mockImplementation((...args) => ({
    en: {
      foo: 'foo',
      bar: { buz: 'buz' }
    },
    ja: {}
  }))
  const writeFiles = {}
  const mockFS = fs as jest.Mocked<typeof fs>
  mockFS.writeFile.mockImplementation((path, data) => {
    writeFiles[path as string] = data
  })

  // run
  const list = await import('../../src/commands/list')
  const cmd = yargs.command(list)
  await new Promise((resolve, reject) => {
    cmd.parse(`list --locale=en --target=./test/fixtures/locales/ja.json`, (err, argv, output) => {
      err ? reject(err) : resolve(output)
    })
  })

  expect(spyLog).toHaveBeenCalledTimes(2)
  expect(spyLog.mock.calls[0]).toEqual([`ja: 'foo' undefined`])
  expect(spyLog.mock.calls[1]).toEqual([`ja: 'bar.buz' undefined`])
  expect(mockFS.writeFile).not.toHaveBeenCalled()
  // NOTE: cannot detect process.exit calling ...
  // expect(process.exit).toHaveBeenCalledWith(5)
})

test('--target-paths option', async () => {
  // setup mocks
  const mockUtils = utils as jest.Mocked<typeof utils>
  mockUtils.getLocaleMessages.mockImplementation((...args) => ({
    en: {
      foo: 'foo',
      bar: { buz: 'buz' }
    },
    ja: {}
  }))
  const writeFiles = {}
  const mockFS = fs as jest.Mocked<typeof fs>
  mockFS.writeFile.mockImplementation((path, data) => {
    writeFiles[path as string] = data
  })

  // run
  const list = await import('../../src/commands/list')
  const cmd = yargs.command(list)
  await new Promise((resolve, reject) => {
    cmd.parse(`list --locale=en \
      --target-paths=./test/fixtures/locales/*.json \
      --filename-match=^([\\w]*)\\.json`, (err, argv, output) => {
      err ? reject(err) : resolve(output)
    })
  })

  expect(spyLog).toHaveBeenCalledTimes(2)
  expect(spyLog.mock.calls[0]).toEqual([`ja: 'foo' undefined`])
  expect(spyLog.mock.calls[1]).toEqual([`ja: 'bar.buz' undefined`])
  expect(mockFS.writeFile).not.toHaveBeenCalled()
  // NOTE: cannot detect process.exit calling ...
  // expect(process.exit).toHaveBeenCalledWith(5)
})

test('--define option', async () => {
  // setup mocks
  const mockUtils = utils as jest.Mocked<typeof utils>
  mockUtils.getLocaleMessages.mockImplementation((...args) => ({
    en: {
      foo: 'foo',
      bar: { buz: 'buz' }
    },
    ja: {}
  }))
  const writeFiles = {}
  const mockFS = fs as jest.Mocked<typeof fs>
  mockFS.writeFile.mockImplementation((p, data) => {
    writeFiles[path.basename(p as string)] = data
  })

  // run
  const list = await import('../../src/commands/list')
  const cmd = yargs.command(list)
  await new Promise((resolve, reject) => {
    cmd.parse(`list --locale=en \
      --target-paths=./test/fixtures/locales/*.json \
      --filename-match=^([\\w]*)\\.json \
      --define`, (err, argv, output) => {
      err ? reject(err) : resolve(output)
    })
  })

  expect(spyLog).toHaveBeenCalledTimes(3)
  expect(spyLog.mock.calls[0]).toEqual([`ja: 'foo' undefined`])
  expect(spyLog.mock.calls[1]).toEqual([`ja: 'bar.buz' undefined`])
  expect(JSON.parse(writeFiles['ja.json'])).toEqual({
    foo: '',
    bar: { buz: '' }
  })
})

test('--indent option', async () => {
  // setup mocks
  const mockUtils = utils as jest.Mocked<typeof utils>
  mockUtils.getLocaleMessages.mockImplementation((...args) => ({
    en: {
      foo: 'foo',
      bar: { buz: 'buz' }
    },
    ja: {}
  }))
  const writeFiles = {}
  const mockFS = fs as jest.Mocked<typeof fs>
  mockFS.writeFile.mockImplementation((p, data) => {
    writeFiles[path.basename(p as string)] = data
  })

  // run
  const list = await import('../../src/commands/list')
  const cmd = yargs.command(list)
  await new Promise((resolve, reject) => {
    cmd.parse(`list --locale=en \
      --target-paths=./test/fixtures/locales/*.json \
      --filename-match=^([\\w]*)\\.json \
      --define --indent=4`, (err, argv, output) => {
      err ? reject(err) : resolve(output)
    })
  })

  expect(writeFiles).toMatchSnapshot()
})
