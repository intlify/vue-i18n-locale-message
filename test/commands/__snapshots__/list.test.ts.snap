// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--indent option 1`] = `
Object {
  "ja.json": "{
    \\"foo\\": \\"\\",
    \\"bar\\": {
        \\"buz\\": \\"\\"
    }
}",
}
`;
